### PR TITLE
[MRG] Build Example Notebook to Walk Through mne-hfo usage

### PR DESCRIPTION
**Pull request by pmyers16**
PR Description
--------------

Closes: #6 
Closes: #9 

Closes: #8 by storing HFO events as `*annotations.tsv` files with proper IO.

This PR's goal is mostly to build a demo notebook to walk through mne-hfo usage. Some modifications and supporting functions needed to be added along the way.

Still in progress.

TODO (I will help w/ this):

- [ ] API to take in two annotations dataframes and find overlapping HFO events
- [x] API to convert from *events.tsv dataframe to *annotations.tsv dataframe
- [ ] API to compute rate of HFOs per channel (per unit of time (e.g. minutes/hrs)).

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [x] All comments resolved
- [x] This is not your own PR
- [x] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/adam2392/mne-hfo/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

File: mne_hfo/io.py [hunk 1]
```diff
@@ -37,6 +37,8 @@ def create_events_df(input: Union[Dict[str, List], mne.io.Raw],
     sfreq : float | None
         The sampling frequency. Only required if the input is
         not a ``mne.io.BaseRaw`` object.
+    event_name: str
+        The name of the event to add to the "trial_type" column
```
Comment (adam2392) [inline mne_hfo/io.py:41]:
```suggestion
        The prefix of the channel event to add to the "trial_type" column (e.g. ``hfo`` for channel ``A1``
        would result in ``hfo_A1``).
```

**Comment by adam2392** [comment]
Nice, it's not bad start. A few comments:

1. Limit the print statements. It renders as a huge file otw.
2. Include a GridSearchCV/RandomGridSearchCV using metric we converge on (see below). Might need to fenangle the Detector implementation though to improve the `predict` and `score` function.
3. Render the simulated part.

To compare HFO detections, what are you using as a metric? E.g. det1 gets HFO (time_x, time_y) and det2 gets HFO (time_a, time_b). What is the proposed overlap function? 

In https://github.com/adam2392/mne-hfo/issues/9, I guess it is just overlap in endpoints counts as +1, no-overlap at all in endpoints counts as 0. Is that correct? Thus the metric in our case is simply a "custom" accuracy implementation.

Are there any weaknesses to this approach? What other metrics can alleviate that weakness?

### Scoring Function to make SearchCV work

If that's the case, it seems that most of traditional metrics (e.g. accuracy), we have to define ourselves using for example, the overlap incidence defined above. Probably more things, such as roc_auc, etc. I can see the `score` function to simply operate with a kwarg input defining the metric (e.g. `'accuracy'`).

Then, we massage the data to go from these HFO events to our criterion. E.g. For overlap, events gets converted from (time_x, time_y) -> {+1, 0}. So for two running lists of HFO events, we get a list of y_true and y_pred. Then it becomes trivial to feed those to sklearn's API. That way, our responsibility is just implementing the logic to get to `y_true` and `y_pred`.

**Comment by adam2392** [comment]
I've added an API to go from `*events.tsv` -> `*annotations.tsv` and corresponding tests for the public functions in ``io.py``. 

Let's get this closed out and refine the example as we develop the API further. Can you do the following:

1. Simplify the jupyter notebook (see comment above) and render the simulation cells as well.
2. Add unit tests for `find_coincident_events` 
3. Update `whats_new.rst` document with a description of what was added here. 

Re-ping me when that's done, and then let's meet to discuss how the algorithms performed on the validation dataset, and how we can start running it on the sickkids dataset.

File: mne_hfo/utils.py [hunk 2]
```diff
@@ -245,6 +258,27 @@ def match_detections(gs_df, dd_df, bn, freq_name=None,
...
 
+def _change_duration_to_offset(dfs, cols):
+    for ind, df in enumerate(dfs):
+        df_col_indices = [df.columns.get_loc(c) for c in cols if c in cols]
+        df['offset'] = df.iloc[:, df_col_indices].sum(axis=1)
+        dfs[ind] = df
+    return dfs
+
+
+def _check_column_types(dfs, cols):
```
Comment (adam2392) [inline mne_hfo/utils.py:269]: What's the point of this function?

File: mne_hfo/utils.py [hunk 3]
```diff
@@ -245,6 +258,27 @@ def match_detections(gs_df, dd_df, bn, freq_name=None,
     return match_df
 
 
+def _change_duration_to_offset(dfs, cols):
```
Comment (adam2392) [inline mne_hfo/utils.py:261]: What's the point of this function?

File: mne_hfo/utils.py [hunk 4]
```diff
@@ -274,3 +308,59 @@ def check_detection_overlap(gs, dd):
         overlap = True
 
     return overlap
+
+
+def find_coincident_events(hfo_dict1, hfo_dict2):
+    """
+    Get a dictionary of hfo events that overlap between two sets.
+
+    Note: Both input dictionaries should come from the same original dataset and therefore contain the same keys
```
Comment (adam2392) [inline mne_hfo/utils.py:317]: Perhaps add an error check at the beginning for this.

File: mne_hfo/utils.py (hunk 2 above)
Comment (adam2392) [reply mne_hfo/utils.py:269]: Okay I think it means you're trying to check numeric for columns. Change the name to something like `_enforce_numeric_cols`, or something like that.

**Review by adam2392 — CHANGES_REQUESTED** [review]
I've added an API to go from `*events.tsv` -> `*annotations.tsv` and corresponding tests for the public functions in ``io.py``. 

Let's get this closed out and refine the example as we develop the API further. Can you do the following:

1. Simplify the jupyter notebook (see comment above) and render the simulation cells as well.
2. Add unit tests for `find_coincident_events` 
3. Update `whats_new.rst` document with a description of what was added here. 

Re-ping me when that's done, and then let's meet to discuss how the algorithms performed on the validation dataset, and how we can start running it on the sickkids dataset.

File: mne_hfo/utils.py (hunk 3 above)
Comment (pmyers16) [reply mne_hfo/utils.py:263]: The current "compare to gold standard" functions use onset and offset columns of a dataframe, but the output dataframes from analysis have onset and duration. This just adds the offset column

Comment (adam2392) [reply mne_hfo/utils.py:263]: Ah I see, rather then pass a list of dataframes(?) and list of columns, perhaps make this more verbose?

def _add_offset_to_df(df):
      # compute offset for one dataframe using onset/duration

Then call it on the outside inside a for loop if you really need to run it on multiple dataframes.

Rn, it seems more general then it should be, and that can lead to errors that are unforseen.

File: docs/whats_new.rst [hunk 5]
```diff
@@ -24,6 +24,7 @@ Authors
 ~~~~~~~
 
 * `Adam Li`_
+* `Patrick Myers`_
```
Comment (adam2392) [inline docs/whats_new.rst:27]: You'll need to add yourself to the `authors.rst` document as well.

**Comment by adam2392** [comment]
Fixed circle CI for you.

**Pull request merged**